core/src/main/java/org/opensearch/sql/expression/operator/predicate/BinaryPredicateOperator.java

```diff
@@ -261,4 +292,31 @@ private static ExprValue lookupTableFunction(ExprValue arg1, ExprValue arg2,
       return table.get(arg2, arg1);
     }
   }
+
+  private static List<Pair<ExprCoreType, ExprCoreType>> permuteTemporalTypesByPairs() {
+    var res = new ArrayList<Pair<ExprCoreType, ExprCoreType>>();
+    var datatypes = List.of(TIME, DATE, DATETIME, TIMESTAMP);
+    datatypes.forEach(left -> {
+      datatypes.forEach(right -> {
+        if (left != right) {
+          res.add(Pair.of(left, right));
+        }
+      });
+    });
+    return res;
+  }
+
+  /**
+   * Convert a temporal ExprValue to LocalDateTime to use in comparison.
+   * A time converted to today's time, a date converted to date's midnight.
+   *
+   * @param value ExprTimeValue/ExprDatetimeValue/ExprDateValue/ExprTimestampValue
+   * @return The input converted/casted to LocalDateTime
+   */
+  private static LocalDateTime convertTemporalToDateTime(ExprValue value) {
```

Comment:
It's odd to special-case one of four possible types. You could override `ExptTimeValue.datetimeValue` to return `timeValue().atDate(LocalDate.now())`. It would be more consistent and simpler.

Reply:
Fixed in 44e24fc.